php/class-wp-js-widget.php

```diff
@@ -459,38 +494,167 @@ protected function render_input_attrs( $input_attrs ) {
+	 *     @type string $type        Form field type, including the values for `input[type]` as well as 'select' and 'textarea'. Optional, defaults to 'text'.
+	 *     @type string $label       Field label, uses schema field `description` by default.
+	 *     @type string $class       Class name. Optional.
+	 *     @type string $placeholder Input placeholder, uses schema field `default` by default. Optional.
+	 *     @type string $min         Minimum, uses schema field `minimum` by default. Optional.
+	 *     @type string $max         Maximum, uses schema field `maximum` by default. Optional.
+	 *     @type string $help        Optional help text that appears below the input.
+	 * }
 	 */
 	protected function render_form_field_template( $args = array() ) {
-		$defaults = array(
-			'name' => '',
-			'label' => '',
+		$item_schema = $this->get_item_schema();
+		$default_input_attrs = array(
 			'type' => 'text',
-			'choices' => array(),
-			'value' => '',
-			'placeholder' => '',
-			'help' => '',
+			'class' => '',
 		);
+
+		if ( ! empty( $args['name'] ) ) {
+			_deprecated_argument( __FUNCTION__, '0.3.0', __( 'The args[name] param is deprecated in favor of args[field].', 'js-widgets' ) );
+
+			if ( empty( $args['field'] ) && ! empty( $item_schema[ $args['name'] ] ) ) {
+				$args['field'] = $args['name'];
+			}
+		}
+
+		$field_name = ! empty( $args['field'] ) ? $args['field'] : null;
+		if ( $field_name && ! empty( $item_schema[ $field_name ] ) ) {
+			$field_schema = $item_schema[ $field_name ];
+			$schema_to_input_attrs_mapping = array(
+				'description' => 'label',
+				'minimum' => 'min',
+				'maximum' => 'max',
+				'default' => 'placeholder',
+			);
+			foreach ( $schema_to_input_attrs_mapping as $schema_key => $input_attr_key ) {
+				if ( isset( $field_schema[ $schema_key ] ) ) {
+					$default_input_attrs[ $input_attr_key ] = $field_schema[ $schema_key ];
+				}
+			}
+			if ( isset( $field_schema['properties']['raw'] ) ) {
+				foreach ( $schema_to_input_attrs_mapping as $schema_key => $input_attr_key ) {
+					if ( isset( $field_schema['properties']['raw'][ $schema_key ] ) ) {
+						$default_input_attrs[ $input_attr_key ] = $field_schema['properties']['raw'][ $schema_key ];
+					}
+				}
+			}
+
+			if ( isset( $field_schema['type'] ) && is_string( $field_schema['type'] ) ) {
+				$schema_type = $field_schema['type'];
+
+				if ( 'boolean' === $schema_type ) {
+					$default_input_attrs['type'] = 'checkbox';
+				} elseif ( 'integer' === $schema_type || 'number' === $schema_type ) {
+					$default_input_attrs['type'] = 'number';
+				} elseif ( 'string' === $schema_type && isset( $field_schema['format'] ) ) {
+					if ( 'uri' === $field_schema['format'] ) {
+						$default_input_attrs['type'] = 'url';
+					} elseif ( 'email' === $field_schema['format'] ) {
+						$default_input_attrs['type'] = 'email';
+					}
```

Comment:
@delawski For the normal usage of `\WP_JS_Widget::render_form_field_template()` this shouldn't happen because it expects in that case that you would supply a `type` that is valid for HTML (and in any case `string` will behave like `text` anyway):

```php
$default_input_attrs = array(
	'type' => 'text',
	'class' => '',
);
```

In your case since you're iterating over the item schema to pass its params directly into `\WP_JS_Widget::render_form_field_template()`, then I think you need to override it.

For example, consider your overridden `\WP_JS_Widget::render_form_field_template()`:

```php
<?php
foreach ( $this->get_item_schema() as $field_name => $field_schema ) {
	$args = array(
		'field' => $field_name,
	);
	if ( isset( $field_schema['input_attrs'] ) ) {
		$args = array_merge( $args, $field_schema['input_attrs'] );
	}
	$this->render_form_field_template( $args );
}
```

And then consider this item schema field:

```php
'layout' => call_user_func( function() {
	$choices = array(
		'3-articles-list' => __( '3 articles list', 'example' ),
		'4-articles'      => __( '4 articles', 'example' ),
		'5-articles'      => __( '5 articles', 'example' ),
		'6-articles'      => __( '6 articles with ad', 'example' ),
	);
	return array(
		'description'   => __( 'Layout', 'example' )
		'type'    => 'select',
		'default' => '3-articles-list',
		'enum'    => array_keys( $choices ),
		'context'     => array( 'view', 'edit', 'embed' ),
		'arg_options' => array(
			'validate_callback' => 'rest_validate_request_arg',
		),
		'input_attrs' => array(
			'type' => 'text', // ☜ ☜ ☜
			'choices' => $choices,
		),
	);
} )
```

Notice in how `type` is overridden in the `input_attrs` schema property. Note also that this `input_attrs` is just something invented here. It's not recognized by the core of the WP REST API or JS Widgets.